src/main/java/htsjdk/samtools/HtsgetBAMFileReader.java

```diff
@@ -0,0 +1,561 @@
+
+    // If true, all SAMRecords are fully decoded as they are read.
+    private boolean mEagerDecode;
+
+    private boolean mCheckCRC;
+
+    private final boolean mUseAsynchronousIO;
+
+    // For error-checking.
+    private ValidationStringency mValidationStringency;
+
+    // For creating BAMRecords
+    private SAMRecordFactory mSamRecordFactory;
+
+    private SamReader mReader;
+
+    // Hold on to all query iterators we've given out so we can close them
+    private final List<CloseableIterator<SAMRecord>> iterators;
+
+    /**
+     * Prepare to read BAM from an htsget source
+     *
+     * @param source               source of bytes.
+     * @param eagerDecode          if true, decode all BAM fields as reading rather than lazily.
+     * @param validationStringency Controls how to handle invalidate reads or header lines.
+     * @param samRecordFactory     SAM record factory
+     * @param useAsynchronousIO    if true, use asynchronous I/O
+     */
+    public HtsgetBAMFileReader(final URI source,
+                               final boolean eagerDecode,
+                               final ValidationStringency validationStringency,
+                               final SAMRecordFactory samRecordFactory,
+                               final boolean useAsynchronousIO) throws IOException {
+        this(source, eagerDecode, validationStringency, samRecordFactory, useAsynchronousIO, BlockGunzipper.getDefaultInflaterFactory());
+    }
+
+    /**
+     * Prepare to read BAM from a htsget source
+     *
+     * @param source               source of bytes.
+     * @param eagerDecode          if true, decode all BAM fields as reading rather than lazily.
+     * @param validationStringency Controls how to handle invalidate reads or header lines.
+     * @param samRecordFactory     SAM record factory
+     * @param useAsynchronousIO    if true, use asynchronous I/O
+     * @param inflaterFactory      InflaterFactory used by BlockCompressedInputStream
+     */
+    public HtsgetBAMFileReader(final URI source,
+                               final boolean eagerDecode,
+                               final ValidationStringency validationStringency,
+                               final SAMRecordFactory samRecordFactory,
+                               final boolean useAsynchronousIO,
+                               final InflaterFactory inflaterFactory) throws IOException {
+        this.mSource = source;
+        this.mEagerDecode = eagerDecode;
+        this.mValidationStringency = validationStringency;
+        this.mSamRecordFactory = samRecordFactory;
+        this.mUseAsynchronousIO = useAsynchronousIO;
+        this.mInflaterFactory = inflaterFactory;
+
+        final HtsgetRequest req = new HtsgetRequest(this.mSource).withDataClass(HtsgetClass.header);
+        // Request only the header and use it to construct a SAMFileHeader for this reader
+        try (final InputStream headerStream = req.getResponse().getDataStream()) {
```

Comment:
I think we have to be checking the response objects for the FORMAT to be sure that they are BAM. (not sure if it matters for the headers, but definitely matters for the data).